File: day21/main.go
package main

import (
	"bufio"
	"fmt"
	"log"
	"os"
	"strings"
)

const (
	puzzleInput = "input.txt"
)

func main() {
	file, err := os.Open(puzzleInput)
	if err != nil {
		log.Fatalln(err)
	}
	defer func() {
		if err := file.Close(); err != nil {
			log.Fatalln(err)
		}
	}()

	var start Pos
	var grid [][]byte
	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		line := scanner.Text()
		if pos := strings.Index(line, "S"); pos >= 0 {
			start = Pos{
				x: pos,
				y: len(grid),
			}
		}
		grid = append(grid, []byte(line))
	}

	if err := scanner.Err(); err != nil {
		log.Fatalln(err)
	}

	height := len(grid)
	width := len(grid[0])
	const target = 26501365
	multiple := target / height
	rem := target % height

	innerEven := 0
	innerOdd := 0
	cornerEven := 0
	cornerOdd := 0

	closedSet := make([]bool, width*height)
	openSet := NewRing[State](width * height)
	closedSet[start.y*width+start.x] = true
	openSet.Write(State{
		pos: start,
		g:   0,
	})
	const p1Target = 64
	const p1TargetIsEven = p1Target%2 == 0
	sum := 0
	for {
		s, ok := openSet.Read()
		if !ok {
			break
		}
		closedSet[s.pos.y*width+s.pos.x] = true
		curIsEven := s.g%2 == 0
		if manhattanDistance(s.pos, start) > rem {
			if curIsEven {
				cornerEven++
			} else {
				cornerOdd++
			}
		} else {
			if curIsEven {
				innerEven++
			} else {
				innerOdd++
			}
		}
		if s.g <= p1Target && curIsEven == p1TargetIsEven {
			sum++
		}
		var neighbors [4]Pos
		n := getNeighbors(grid, width, height, s.pos, neighbors[:])
		for _, i := range neighbors[:n] {
			key := i.y*width + i.x
			if closedSet[key] {
				continue
			}
			closedSet[key] = true
			openSet.Write(State{
				pos: i,
				g:   s.g + 1,
			})
		}
	}
	fmt.Println("Part 1:", sum)

	if height != width {
		log.Fatalln("Grid is not square")
	}
	if height%2 != 1 || start.y != (height-1)/2 || start.x != start.y {
		log.Fatalln("Start is not centered")
	}

	const targetIsEven = target%2 == 0
	multipleIsEven := multiple%2 == 0
	multiple1 := multiple + 1
	outerMultiple := multiple1 * multiple1
	innerMultiple := multiple * multiple

	outerDiamond := innerOdd
	innerDiamond := innerEven
	outerCorner := cornerOdd
	innerCorner := cornerEven
	if targetIsEven == multipleIsEven {
		outerDiamond, innerDiamond = innerDiamond, outerDiamond
		outerCorner, innerCorner = innerCorner, outerCorner
	}
	fmt.Println("Part 2:", outerMultiple*outerDiamond+innerMultiple*innerDiamond+(outerMultiple-multiple1)*outerCorner+(innerMultiple+multiple)*innerCorner)
}

func manhattanDistance(a, b Pos) int {
	return abs(a.x-b.x) + abs(a.y-b.y)
}

func abs(a int) int {
	if a < 0 {
		return -a
	}
	return a
}

type (
	State struct {
		pos Pos
		g   int
	}

	Pos struct {
		x, y int
	}
)

func getNeighbors(grid [][]byte, w, h int, pos Pos, res []Pos) int {
	count := 0
	{
		v := pos
		v.y--
		if inBounds(v, w, h) && grid[v.y][v.x] != '#' {
			res[count] = v
			count++
		}
	}
	{
		v := pos
		v.x--
		if inBounds(v, w, h) && grid[v.y][v.x] != '#' {
			res[count] = v
			count++
		}
	}
	{
		v := pos
		v.y++
		if inBounds(v, w, h) && grid[v.y][v.x] != '#' {
			res[count] = v
			count++
		}
	}
	{
		v := pos
		v.x++
		if inBounds(v, w, h) && grid[v.y][v.x] != '#' {
			res[count] = v
			count++
		}
	}
	return count
}

func inBounds(pos Pos, w, h int) bool {
	return pos.x >= 0 && pos.y >= 0 && pos.x < w && pos.y < h
}

type (
	Ring[T any] struct {
		buf []T
		r   int
		w   int
	}
)

func NewRing[T any](size int) *Ring[T] {
	if size < 2 {
		size = 2
	}
	return &Ring[T]{
		buf: make([]T, size),
		r:   0,
		w:   0,
	}
}

func (b *Ring[T]) resize() {
	next := make([]T, len(b.buf)*2)
	if b.r == b.w {
		b.w = 0
	} else if b.r < b.w {
		b.w = copy(next, b.buf[b.r:b.w])
	} else {
		p := copy(next, b.buf[b.r:])
		q := 0
		if b.w > 0 {
			q = copy(next[p:], b.buf[:b.w])
		}
		b.w = p + q
	}
	b.buf = next
	b.r = 0
}

func (b *Ring[T]) Write(m T) {
	next := (b.w + 1) % len(b.buf)
	if next == b.r {
		b.resize()
		b.Write(m)
		return
	}
	b.buf[b.w] = m
	b.w = next
}

func (b *Ring[T]) Read() (T, bool) {
	if b.r == b.w {
		var v T
		return v, false
	}
	next := (b.r + 1) % len(b.buf)
	m := b.buf[b.r]
	b.r = next
	return m, true
}

func (b *Ring[T]) Peek() (T, bool) {
	if b.r == b.w {
		var v T
		return v, false
	}
	m := b.buf[b.r]
	return m, true
}

File: day21/src/main.rs
use std::collections::VecDeque;
use std::fs::File;
use std::io::{prelude::*, BufReader};

const PUZZLEINPUT: &str = "input.txt";

const TARGET: usize = 26501365;
const TARGET_IS_EVEN: bool = TARGET % 2 == 0;
const P1_TARGET: usize = 64;
const P1_TARGET_IS_EVEN: bool = P1_TARGET % 2 == 0;

fn main() -> Result<(), Box<dyn std::error::Error>> {
    let file = File::open(PUZZLEINPUT)?;
    let reader = BufReader::new(file);

    let mut start = Pos { x: 0, y: 0 };
    let mut grid = Vec::new();
    for line in reader.lines() {
        let line = line?.into_bytes();
        if let Some(x) = line.iter().position(|&v| v == b'S') {
            start = Pos { x, y: grid.len() };
        }
        grid.push(line);
    }

    let h = grid.len();
    let w = grid[0].len();
    let multiple = TARGET / h;
    let rem = TARGET % h;

    let graph = GridGraph { w, h, grid };
    let mut closed_set = ClosedSetVec {
        w,
        vec: vec![false; w * h],
        rem,
        start,
        sum: 0,
        inner_even: 0,
        inner_odd: 0,
        corner_even: 0,
        corner_odd: 0,
    };
    bfs(start, &graph, &mut closed_set);
    println!("Part 1: {}", closed_set.sum);

    if h != w {
        return Err("Grid is not square".into());
    }
    if h % 2 != 1 || start.y != (h - 1) / 2 || start.x != start.y {
        return Err("Start is not centered".into());
    }

    let multiple_is_even = multiple % 2 == 0;
    let multiple1 = multiple + 1;
    let outer_multiple = multiple1 * multiple1;
    let inner_multiple = multiple * multiple;

    let (outer_diamond, inner_diamond, outer_corner, inner_corner) =
        if TARGET_IS_EVEN == multiple_is_even {
            (
                closed_set.inner_even,
                closed_set.inner_odd,
                closed_set.corner_even,
                closed_set.corner_odd,
            )
        } else {
            (
                closed_set.inner_odd,
                closed_set.inner_even,
                closed_set.corner_odd,
                closed_set.corner_even,
            )
        };

    println!(
        "Part 2: {}",
        outer_multiple * outer_diamond
            + inner_multiple * inner_diamond
            + (outer_multiple - multiple1) * outer_corner
            + (inner_multiple + multiple) * inner_corner
    );

    Ok(())
}

#[derive(Eq, PartialEq, Clone, Copy)]
struct Pos {
    x: usize,
    y: usize,
}

fn in_bounds(a: Pos, w: usize, h: usize) -> bool {
    a.x < w && a.y < h
}

fn manhattan_distance(a: &Pos, b: &Pos) -> usize {
    a.x.abs_diff(b.x) + a.y.abs_diff(b.y)
}

struct GridGraph {
    w: usize,
    h: usize,
    grid: Vec<Vec<u8>>,
}

impl Graph<Pos> for GridGraph {
    fn is_goal(&self, _a: &Pos) -> bool {
        false
    }

    fn get_edges(&self, &a: &Pos, res: &mut Vec<Pos>) {
        let mut s = a;
        s.y = s.y.wrapping_sub(1);
        if in_bounds(s, self.w, self.h) && self.grid[s.y][s.x] != b'#' {
            res.push(s);
        }
        let mut s = a;
        s.x += 1;
        if in_bounds(s, self.w, self.h) && self.grid[s.y][s.x] != b'#' {
            res.push(s);
        }
        let mut s = a;
        s.y += 1;
        if in_bounds(s, self.w, self.h) && self.grid[s.y][s.x] != b'#' {
            res.push(s);
        }
        let mut s = a;
        s.x = s.x.wrapping_sub(1);
        if in_bounds(s, self.w, self.h) && self.grid[s.y][s.x] != b'#' {
            res.push(s);
        }
    }
}

struct ClosedSetVec {
    w: usize,
    vec: Vec<bool>,
    rem: usize,
    start: Pos,
    sum: usize,
    inner_even: usize,
    inner_odd: usize,
    corner_even: usize,
    corner_odd: usize,
}

impl ClosedSet<Pos> for ClosedSetVec {
    fn insert(&mut self, a: &Pos, g: usize) {
        self.vec[self.w * a.y + a.x] = true;
        let cur_is_even = g % 2 == 0;
        if g <= P1_TARGET && cur_is_even == P1_TARGET_IS_EVEN {
            self.sum += 1;
        }
        if manhattan_distance(&self.start, a) > self.rem {
            if cur_is_even {
                self.corner_even += 1;
            } else {
                self.corner_odd += 1;
            }
        } else {
            if cur_is_even {
                self.inner_even += 1;
            } else {
                self.inner_odd += 1;
            }
        }
    }

    fn insert_parent(&mut self, _a: &Pos, _g: usize, _p: &Pos) {}

    fn contains(&self, a: &Pos) -> bool {
        self.vec[self.w * a.y + a.x]
    }
}

#[derive(Eq, PartialEq)]
struct Node<T> {
    value: T,
    g: usize,
}

trait Graph<T> {
    fn is_goal(&self, a: &T) -> bool;
    fn get_edges(&self, a: &T, res: &mut Vec<T>);
}

trait ClosedSet<T> {
    fn insert(&mut self, a: &T, g: usize);
    fn insert_parent(&mut self, a: &T, g: usize, p: &T);
    fn contains(&self, a: &T) -> bool;
}

fn bfs<T: Eq>(
    start: T,
    graph: &impl Graph<T>,
    closed_set: &mut impl ClosedSet<T>,
) -> Option<(T, usize)> {
    let mut open_set = VecDeque::new();
    closed_set.insert(&start, 0);
    open_set.push_back(Node { value: start, g: 0 });
    let mut edges = Vec::new();
    loop {
        let current = match open_set.pop_front() {
            Some(v) => v,
            None => return None,
        };
        if graph.is_goal(&current.value) {
            return Some((current.value, current.g));
        }
        graph.get_edges(&current.value, &mut edges);
        while let Some(value) = edges.pop() {
            if closed_set.contains(&value) {
                continue;
            }
            let g = current.g + 1;
            closed_set.insert(&value, g);
            closed_set.insert_parent(&value, g, &current.value);
            open_set.push_back(Node { value, g });
        }
    }
}
